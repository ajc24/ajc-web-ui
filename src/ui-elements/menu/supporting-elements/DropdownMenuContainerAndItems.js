import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import SmallRightArrowIcon from '../../icons/SmallRightArrowIcon';
import '../../../css/common.css';
import '../../../css/menu.css';

class DropdownMenuContainerAndItems extends React.Component {
  constructor(props) {
    super(props);
    this.setDropdownItemsListXPosition = this.setDropdownItemsListXPosition.bind(this);
    this.setDropdownItemsListYPosition = this.setDropdownItemsListYPosition.bind(this);
  }

  componentDidMount() {
    setTimeout(() => {
      this.setDropdownItemsListYPosition();
      this.setDropdownItemsListXPosition();
      window.addEventListener('resize', this.setDropdownItemsListXPosition, { capture: true });
    }, 10);
  }

  componentWillUnmount() {
    /* Remove the dynamically added resize event when the component is unmounted */
    window.removeEventListener('resize', this.setDropdownItemsListXPosition, { capture: true });
  }

  /**
   * Sets the X position of the dropdown menu items list container
   * The X position is always re-adjustable since screen resizes will affect the X position of the container
   */
  setDropdownItemsListXPosition() {
    /* Find the boundaries of the main menu item for this component */
    const mainMenuItemElement = document.querySelector(`div[id="${this.props.parentId}"]`);
    const mainMenuItemBoundary = mainMenuItemElement.getBoundingClientRect();
    /* Find the dropdown items list root element and set the width */
    const itemsListRootElement = document.querySelector(`div[id="${this.props.id}"]`);
    const itemsListRootBoundary = itemsListRootElement.getBoundingClientRect();
    if (itemsListRootBoundary.width < mainMenuItemBoundary.width) {
      itemsListRootElement.style.width = `${mainMenuItemBoundary.width}px`;
    } else {
      itemsListRootElement.style.width = `${itemsListRootBoundary.width}px`;
    }
    itemsListRootElement.style.minWidth = itemsListRootElement.style.width;
    /* Determine if the leftmost position of the dropdown menu items container needs to be adjusted to fit the screen size */
    let leftmostPosition = mainMenuItemBoundary.left;
    if (mainMenuItemBoundary.left + itemsListRootBoundary.width > window.innerWidth) {
      leftmostPosition = window.innerWidth - itemsListRootBoundary.width;
    }
    itemsListRootElement.style.left = `${leftmostPosition}px`;
  }

  /**
   * Sets the Y position of the dropdown menu items list container
   * This only needs to be set once since screen resizes will only mainly affect the X axis positioning of the container
   */
  setDropdownItemsListYPosition() {
    /* Ensure the window is scrolled to the top before marking the position of the dropdown items list */
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    });
    /* Find the boundaries of the main menu item for this component */
    const mainMenuItemElement = document.querySelector(`div[id="${this.props.parentId}"]`);
    const mainMenuItemBottomPos = mainMenuItemElement.getBoundingClientRect().bottom;
    /* Find the dropdown items list root element and set the top position */
    const itemsListRootElement = document.querySelector(`div[id="${this.props.id}"]`);
    itemsListRootElement.style.top = `${mainMenuItemBottomPos}px`;
  }

  render() {
    let rootStyling = 'ajc-flex-column ajc-dropdown-menu';
    let dropdownMenuItemStyling = 'ajc-flex-row ajc-dropdown-menu-item ajc-font-menu';
    if (this.props.colour === 'red') {
      rootStyling += ' ajc-background-red';
      dropdownMenuItemStyling += ' ajc-menu-item-red';
    } else {
      rootStyling += ' ajc-background-grey-2';
      dropdownMenuItemStyling += ' ajc-menu-item-grey';
    }
    if (this.props.isDisplayed === true) {
      rootStyling += ' ajc-dropdown-menu-visible ajc-visibility-visible';
    } else {
      rootStyling += ' ajc-dropdown-menu-hidden ajc-visibility-hidden';
    }
    return (
      <div id={this.props.id} className={rootStyling}>
        {
          this.props.dropdownMenuItemsList.map((menuItem, index) => {
            return <Link to={menuItem.route} key={`${this.props.id}--key-menu-item-${index}`} className={dropdownMenuItemStyling} id={`${this.props.id}--dropdown-menu-item-${index}`} tabIndex={0}>
              &nbsp;&nbsp;{menuItem.title}&nbsp;&nbsp;
              <SmallRightArrowIcon colour="white" />
              &nbsp;&nbsp;
            </Link>
          })
        }
      </div>
    );
  }
}
DropdownMenuContainerAndItems.propTypes = {
  /** The components unique identifier. */
  id: PropTypes.string.isRequired,
  /** The background colour for the container. By default this is set to grey. */
  colour: PropTypes.oneOf([ 'grey', 'red' ]),
  /** The unique identifier of this components parent DropdownMenu component. */
  parentId: PropTypes.string.isRequired,
  /** Specifies whether the dropdown menu container is visible or hidden in the UI. */
  isDisplayed: PropTypes.bool.isRequired,
  /** The dropdown menu data which defines the list of dropdown menu items and defines the route and title for each individual item. */
  dropdownMenuItemsList: PropTypes.arrayOf(PropTypes.shape({
    route: PropTypes.string,
    title: PropTypes.string
  })).isRequired,
};
DropdownMenuContainerAndItems.defaultProps = {
  colour: 'grey'
};
export default DropdownMenuContainerAndItems;
