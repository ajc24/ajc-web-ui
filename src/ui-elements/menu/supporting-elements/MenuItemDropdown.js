import React from 'react';
import PropTypes from 'prop-types';
import DropdownMenuContainerAndItems from './DropdownMenuContainerAndItems';
import SmallDownArrowIcon from '../../icons/SmallDownArrowIcon';
import '../../../css/common.css';
import '../../../css/menu.css';

class MenuItemDropdown extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isExpanded: false
    };
    /* Bind all relevant functions to the component */
    this.handleMenuBarItemKeyPress = this.handleMenuBarItemKeyPress.bind(this);
    this.hideDropdownMenu = this.hideDropdownMenu.bind(this);
    this.showDropdownMenu = this.showDropdownMenu.bind(this);
  }

  /**
   * Handles key press events on the dropdown menu bar item
   * @param {Event} event 
   */
  handleMenuBarItemKeyPress(event) {
    if (event.key === 'Enter'){
      const menuBarItemContent = document.querySelector(`div[id="${this.props.id}"] > div[class*="ajc-menu-item-content"]`);
      menuBarItemContent.click();
    }
  }

  /**
   * Hides the dropdown menu from the UI
   * @param {Event}
   */
  hideDropdownMenu(event) {
    event.preventDefault();
    if (event.target.classList.contains('ajc-menu-item-content') && event.target.getAttribute('data-expanded') === 'true') {
      /* Clicks on expanded dropdown menu items are registered twice (which closes and then re-opens the dropdown menu again) so ensure only one click is registered */
      event.stopPropagation();
    }
    if (event.target.getAttribute('id') !== `${this.props.id}--content-parent`) {
      /* If the user has clicked any item other than a dropdown menu item then collapse the dropdown menu */
      const newState = Object.assign({}, this.state, {
        isExpanded: false
      });
      this.setState(newState, () => {
        document.removeEventListener('click', this.hideDropdownMenu, { capture: true });
      });
    }
    /* If a dropdown menu item has been clicked then ensure the click event takes place - event.preventDefault() will have otherwise prevented this */
    if (event.target.getAttribute('id') && event.target.getAttribute('id').includes('--dropdown-menu-item-')) {
      event.target.click();
    }
    /* If a single standalone menu item has been clicked then ensure the click event takes place - event.preventDefault() will have otherwise prevented this */
    if (event.target.classList && event.target.classList.length === 4 && event.target.classList[1] === 'ajc-menu-item-content'
      && event.target.classList[2] === 'ajc-font-menu') {
        event.target.click();
        document.activeElement.blur();
    }
  }

  /**
   * Displays the dropdown menu in the UI and adds a click listener to auto-hide the menu again
   * @param {Event} event 
   */
  showDropdownMenu(event) {
    event.preventDefault();
    const newState = Object.assign({}, this.state, {
      isExpanded: true
    });
    this.setState(newState, () => {
      document.addEventListener('click', this.hideDropdownMenu, { capture: true });
    });
  }

  render() {
    /* Set the styling for the component */
    const rootStyling = 'ajc-flex-row ajc-justify-left ajc-menu-item';
    let menuItemStyling = 'ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white';
    if (this.props.colour === 'red') {
      menuItemStyling += ' ajc-menu-item-red';
    } else {
      menuItemStyling += ' ajc-menu-item-grey';
    }
    /* Set the menu item component to be displayed */
    let onClickFunctionality;
    if (this.state.isExpanded === true) {
      onClickFunctionality = this.hideDropdownMenu;
    } else {
      onClickFunctionality = this.showDropdownMenu;
    }
    return (
      <div id={`${this.props.id}`} className={rootStyling}>
        <div className={menuItemStyling} onClick={onClickFunctionality} onKeyPress={this.handleMenuBarItemKeyPress} data-expanded={`${this.state.isExpanded}`} tabIndex={0}>
          &nbsp;&nbsp;{this.props.title}&nbsp;&nbsp;
          <SmallDownArrowIcon colour="white" />
          &nbsp;&nbsp;
        </div>
        <DropdownMenuContainerAndItems id={`${this.props.id}--content-parent`} colour={this.props.colour} parentId={this.props.id} isDisplayed={this.state.isExpanded} dropdownMenuItemsList={this.props.dropdownMenuItemsList} />
      </div>
    );
  }
}
MenuItemDropdown.propTypes = {
  /** The components unique identifier. */
  id: PropTypes.string.isRequired,
  /** The hover colour for the dropdown menu element. By default this is set to grey. */
  colour: PropTypes.oneOf([ 'grey', 'red' ]),
  /** The title for the dropdown menu being displayed. */
  title: PropTypes.string.isRequired,
  /** The dropdown menu data which defines the list of dropdown menu items and defines the route and title for each individual item */
  dropdownMenuItemsList: PropTypes.arrayOf(PropTypes.shape({
    route: PropTypes.string,
    title: PropTypes.string
  })).isRequired,
};
MenuItemDropdown.defaultProps = {
  colour: 'grey'
};
export default MenuItemDropdown;
