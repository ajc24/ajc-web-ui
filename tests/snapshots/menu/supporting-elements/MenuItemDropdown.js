import 'jsdom-global/register';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { TestDev } from 'ajc-jest-enzyme';
import MenuItemDropdown from '../../../../src/ui-elements/menu/supporting-elements/MenuItemDropdown';
import DropdownMenuContainerAndItems from '../../../../src/ui-elements/menu/supporting-elements/DropdownMenuContainerAndItems';

describe('MenuItemDropdown', () => {
  /* Test data */
  let componentDidMountSpy;
  const testDropdownMenuItemsList = [
    {
      route: '/test-menu-item-1',
      title: 'Dropdown Menu Item 1',
    },
    {
      route: '/test-menu-item-2',
      title: 'Dropdown Menu Item 2',
    },
  ];

  beforeAll(() => {
    componentDidMountSpy = jest
      .spyOn(DropdownMenuContainerAndItems.prototype, 'componentDidMount')
      .mockImplementation(() => {});
  });

  afterAll(() => {
    componentDidMountSpy.mockRestore();
  });

  describe('Default props and rendering', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <div role="navigation">
          <BrowserRouter>
            <MenuItemDropdown id="test-id" title="Dropdown Menu" dropdownMenuItemsList={testDropdownMenuItemsList} />
          </BrowserRouter>
        </div>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });

  describe('Transferred props and rendering - Component in a red colour', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <div role="navigation">
          <BrowserRouter>
            <MenuItemDropdown id="test-id" title="Dropdown Menu" colour="red" dropdownMenuItemsList={testDropdownMenuItemsList} />
          </BrowserRouter>
        </div>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });
});
