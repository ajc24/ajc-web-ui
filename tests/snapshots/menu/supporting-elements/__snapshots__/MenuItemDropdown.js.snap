// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemDropdown Default props and rendering verifies the snapshot for the component 1`] = `
<div
  role="navigation"
>
  <div
    className="ajc-flex-row ajc-justify-left ajc-menu-item"
    id="test-id"
  >
    <div
      className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-grey"
      data-expanded="false"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
    >
        
      Dropdown Menu
        
      <div
        className="ajc-flex-row ajc-justify-left"
        role="presentation"
      >
        <div
          className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
        />
      </div>
        
    </div>
    <div
      className="ajc-flex-column ajc-dropdown-menu ajc-background-grey-2 ajc-dropdown-menu-hidden ajc-visibility-hidden"
      id="test-id--content-parent"
    >
      <a
        className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
        href="/test-menu-item-1"
        id="test-id--content-parent--dropdown-menu-item-0"
        onClick={[Function]}
        tabIndex={0}
      >
          
        Dropdown Menu Item 1
          
        <div
          className="ajc-flex-row ajc-justify-left"
          role="presentation"
        >
          <div
            className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
          />
        </div>
          
      </a>
      <a
        className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
        href="/test-menu-item-2"
        id="test-id--content-parent--dropdown-menu-item-1"
        onClick={[Function]}
        tabIndex={0}
      >
          
        Dropdown Menu Item 2
          
        <div
          className="ajc-flex-row ajc-justify-left"
          role="presentation"
        >
          <div
            className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
          />
        </div>
          
      </a>
    </div>
  </div>
</div>
`;

exports[`MenuItemDropdown Transferred props and rendering - Component in a red colour verifies the snapshot for the component 1`] = `
<div
  role="navigation"
>
  <div
    className="ajc-flex-row ajc-justify-left ajc-menu-item"
    id="test-id"
  >
    <div
      className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-red"
      data-expanded="false"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
    >
        
      Dropdown Menu
        
      <div
        className="ajc-flex-row ajc-justify-left"
        role="presentation"
      >
        <div
          className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
        />
      </div>
        
    </div>
    <div
      className="ajc-flex-column ajc-dropdown-menu ajc-background-red ajc-dropdown-menu-hidden ajc-visibility-hidden"
      id="test-id--content-parent"
    >
      <a
        className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
        href="/test-menu-item-1"
        id="test-id--content-parent--dropdown-menu-item-0"
        onClick={[Function]}
        tabIndex={0}
      >
          
        Dropdown Menu Item 1
          
        <div
          className="ajc-flex-row ajc-justify-left"
          role="presentation"
        >
          <div
            className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
          />
        </div>
          
      </a>
      <a
        className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
        href="/test-menu-item-2"
        id="test-id--content-parent--dropdown-menu-item-1"
        onClick={[Function]}
        tabIndex={0}
      >
          
        Dropdown Menu Item 2
          
        <div
          className="ajc-flex-row ajc-justify-left"
          role="presentation"
        >
          <div
            className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
          />
        </div>
          
      </a>
    </div>
  </div>
</div>
`;
